src/useField.js

```diff
@@ -0,0 +1,205 @@
+// @flow
+import * as React from 'react'
+import { fieldSubscriptionItems } from 'final-form'
+import flattenSubscription from './flattenSubscription'
+import type { FieldSubscription, FieldState, FormApi } from 'final-form'
+import type { UseFieldConfig, FieldInputProps, FieldRenderProps } from './types'
+import isReactNative from './isReactNative'
+import getValue from './getValue'
+import ReactFinalFormContext from './context'
+
+const all: FieldSubscription = fieldSubscriptionItems.reduce((result, key) => {
+  result[key] = true
+  return result
+}, {})
+
+export const defaultFormat = (value: ?any, name: string) =>
+  value === undefined ? '' : value
+export const defaultParse = (value: ?any, name: string) =>
+  value === '' ? undefined : value
+
+const useField = (
+  name: string,
+  {
+    allowNull,
+    component,
+    defaultValue,
+    format,
+    formatOnBlur,
+    initialValue,
+    isEqual,
+    multiple,
+    parse,
+    subscription,
+    type,
+    validate,
+    validateFields,
+    value: _value
+  }: UseFieldConfig = {}
+): FieldRenderProps => {
+  const reactFinalForm: ?FormApi = React.useContext(ReactFinalFormContext)
+  // istanbul ignore next
+  if (process.env.NODE_ENV !== 'production' && !reactFinalForm) {
+    console.error(
+      'Warning: useField must be used inside of a ReactFinalForm component'
+    )
+  }
+
+  // keep ref to most recent copy of validate function
+  const validateRef = React.useRef(validate)
+  validateRef.current = validate
+
+  const register = (callback: FieldState => void) =>
+    reactFinalForm
+      ? reactFinalForm.registerField(name, callback, subscription || all, {
+          defaultValue,
+          getValidator: () => validateRef.current,
+          initialValue,
+          isEqual,
+          validateFields
+        })
+      : () => {}
+
+  const firstRender = React.useRef(true)
+
+  // synchronously register and unregister to query field state for our subscription on first render
+  const [state, setState] = React.useState<FieldState>(
+    (): FieldState => {
+      let initialState: FieldState = {}
+      register(state => {
+        initialState = state
+      })()
+      return initialState
+    }
+  )
+
+  const flattenedSubscription = flattenSubscription(subscription || all)
+  // useDependenciesDebugger({
+  //   name,
+  //   defaultValue,
+  //   validate,
+  //   initialValue,
+  //   isEqual,
+  //   validateFields,
+  //   reactFinalForm
+  // })
+  React.useEffect(
+    () =>
+      register(state => {
+        if (firstRender.current) {
+          firstRender.current = false
+        } else {
+          setState(state)
+        }
+      }),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [
+      name,
+      defaultValue,
+      // If we want to allow inline fat-arrow field-level validation functions, we
+      // cannot reregister field every time validate function !==.
+      // validate,
```

Comment:
is this comment still relevant - https://github.com/final-form/react-final-form/pull/467/files#diff-04c6e90faac2675aa89e2176d2eec7d8R574 ? by keeping validate in ref (which you do), you can guarantee that validate can be inlined and it can also be conditional and still wont cause reregister

Reply:
I left the `// validate` commented to show future me / others that including it was, indeed, considered, but a decision was intentionally made to not include it.

If you mean something else, say it again.

Reply:
Just a question for now - seems slightly weird that `validate` gets special cased like this. Is there no other prop that should get such treatment as well? (or none should?)